src/backend/optimizer/util/plancat.c

```diff
@@ -510,7 +518,40 @@ cdb_estimate_rel_size(RelOptInfo   *relOptInfo,
 		 * to derive curpages, else use the default value.
 		 */
 		if (gp_enable_relsize_collection)
-			curpages = cdbRelMaxSegSize(rel) / BLCKSZ;
+		{
+			if (RelationIsAppendOptimized(rel))
+			{
+				FileSegTotals *totals;
+				if (RelationIsAoRows(rel))
+					totals = GetSegFilesTotalsCluster(rel);
+				else
+					totals = GetAOCSSegFilesTotalsCluster(rel);
```

Comment:
The FileSegTotals structure is small, so it can be allocated on stack. What about getting rid of palloc, Assert(totals) and pfree?
```suggestion
				FileSegTotals totals;
				if (RelationIsAoRows(rel))
					GetSegFilesTotalsCluster(rel, &totals);
				else
					GetAOCSSegFilesTotalsCluster(rel, &totals);
```

Reply:
I thought about it also.
But 
 - similar segment local functions like `extern FileSegTotals *GetSegFilesTotals(Relation parentrel, Snapshot appendOnlyMetaDataSnapshot);` do it in this way (palloc). 
 - And it will complicate the logic inside `GetAOCSSegFilesTotalsCluster()` and `GetSegFilesTotalsCluster()`, as the structure is filled inside PG_TRY, so if the struct is on stack, we'll need to handle the stack issues related to PG_TRY. 
 
 So I decided to stay with palloc'ed struct.

Reply:
May be palloc will be better in the case of `GetSegFilesTotals`, but in our case the code become more complex and palloc is not so fast as allocation on stack. There will be no need to handle the stack issues related to PG_TRY, because the variable will not be used in PG_CATCH.

Reply:
Updated